### PR TITLE
Feature/check svg file type

### PR DESCRIPTION
**Pull request by szabozoltan69**
Validating file uploads.
Also fixes https://github.com/IFRCGo/go-api/issues/2077 (which issue was a blocker of proper testing).

### CODE REVIEW

File: per/serializers.py [hunk 1]
```diff
@@ -541,6 +555,8 @@ class Meta:
         read_only_fields = ("created_by",)
 
     def create(self, validated_data):
+        if "file" in validated_data:
+            check(validated_data["file"])
```
Comment (thenav56) [inline per/serializers.py:559]: Same as other, We will need functional validation

File: per/serializers.py [hunk 2]
```diff
@@ -995,13 +1011,14 @@ def validate(self, data):
             )
         return data
 
-
     def create(self, validated_data):
+        if "file" in validated_data:
+            check(validated_data["file"])
```
Comment (thenav56) [inline per/serializers.py:1016]: For create, We can ignore "file" in validated_data check as the `file is a required field in PerDocumentUpload.

File: per/serializers.py [hunk 3]
```diff
@@ -222,6 +223,19 @@ class Meta:
...
+class NiceDocumentSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = NiceDocument
+        fields = "__all__"
+
+    def create(self, validated_data):
+        if "document" in validated_data:
+            check(validated_data["document"])
+        # validated_data["created_by"] = self.context["request"].user  # FIXME: this field would be fine.
```
Comment (thenav56) [inline per/serializers.py:235]:
```suggestion
```

File: flash_update/serializers.py [hunk 4]
```diff
@@ -59,6 +58,8 @@ class Meta:
         read_only_fields = ('created_by',)
 
     def create(self, validated_data):
+        if 'file' in validated_data:
+            check(validated_data['file'])
```
Comment (thenav56) [inline flash_update/serializers.py:62]: For create, We can ignore "file" in validated_data check as the `file is a required field in FlashGraphicMap.

File: dref/serializers.py [hunk 5]
```diff
@@ -66,6 +67,8 @@ class Meta:
         read_only_fields = ("created_by",)
 
     def create(self, validated_data):
+        if "file" in validated_data:
+            check(validated_data["file"])
```
Comment (thenav56) [inline dref/serializers.py:71]: For create, We can ignore "file" in validated_data check as the `file is a required field in DrefFile.

To make sure it is validated for update as well, a function validate_ will be easier.

```python
def validate_file(self, file):
    check(file)
    return file
```

File: dref/serializers.py [hunk 6]
```diff
@@ -471,6 +474,8 @@ def validate_operation_timeframe(self, operation_timeframe):
         return operation_timeframe
 
     def create(self, validated_data):
+        if "budget_file_preview" in validated_data:
+            check(validated_data["budget_file_preview"])
```
Comment (thenav56) [inline dref/serializers.py:478]: Same here, we will need functional validate instead

File: dref/serializers.py [hunk 7]
```diff
@@ -617,6 +622,8 @@ def get_total_timeframe(self, start_date, end_date):
         return None
 
     def create(self, validated_data):
+        if "budget_file_preview" in validated_data:
```
Comment (thenav56) [inline dref/serializers.py:625]: Same here, we will need functional validate instead

File: dref/serializers.py [hunk 8]
```diff
@@ -927,6 +934,9 @@ def create(self, validated_data):
         # here check if there is operational update for corresponding dref
         # if yes copy from the latest operational update
         # else copy from dref
+
+        if "financial_report_preview" in validated_data:
```
Comment (thenav56) [inline dref/serializers.py:938]: Same here, we will need functional validate instead

File: databank/serializers.py [hunk 9]
```diff
@@ -83,6 +84,11 @@ class Meta:
         model = KeyDocument
         exclude = ('overview',)
 
+    def create(self, validated_data):
+        if 'file' in validated_data:
+            check(validated_data['file'])
```
Comment (thenav56) [inline databank/serializers.py:89]: Same as other, We will need functional validation

File: api/models.py [hunk 10]
```diff
@@ -647,7 +647,7 @@ class Meta:
         verbose_name_plural = _('region snippets')
 
     def __str__(self):
-        return self.snippet
+        return (self.snippet[:80] if self.snippet else None) or (self.image.url if getattr(self, 'image') else '')
```
Comment (thenav56) [inline api/models.py:650]: Without default value in `getattr`, it throws AttributeError error
```suggestion
        return (
            self.snippet[:80] if self.snippet else None
        ) or (
            self.image.url
            if getattr(self, 'image', None) and self.image.url
            else ''
        )
```

File: api/models.py [hunk 11]
```diff
@@ -724,7 +724,7 @@ class Meta:
         verbose_name_plural = _('country snippets')
 
     def __str__(self):
-        return self.snippet
+        return (self.snippet[:80] if self.snippet else None) or (self.image.url if getattr(self, 'image') else '')
```
Comment (thenav56) [inline api/models.py:727]: Same as before

File: api/models.py [hunk 12]
```diff
@@ -1020,7 +1020,7 @@ class Meta:
         verbose_name_plural = _('snippets')
 
     def __str__(self):
-        return self.snippet if self.snippet else ''
+        return (self.snippet[:80] if self.snippet else None) or (self.image.url if getattr(self, 'image') else '')
```
Comment (thenav56) [inline api/models.py:1023]: Same as before

File: utils/file_check.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    Raises:
+        ValidationError: If the file type is disallowed.
+    """
+
+    if file:
+        mime = Magic(mime=True)
+        file_type = mime.from_buffer(file.read(1024))
+        file.seek(0)  # Read a portion for analysis, then reset file pointer
+        if file_type == "image/svg+xml":
+            raise ValidationError("SVG image upload is not allowed. Please use another image type.")
```
Comment (thenav56) [inline utils/file_check.py:19]: We will need to use a different name for this function.

Maybe, `validate_allowed_file_type`

Also, let's add a option for bool response
```suggestion
def validate_allowed_file_type(file, raise_exception=True) -> bool:
    """
    This function validates a file object based on a disallowed mimetype.
    Arg:
        file: The file object to validate.
    Raises:
        ValidationError: If the file type is disallowed.
    """

    if file:
        mime = Magic(mime=True)
        file_type = mime.from_buffer(file.read(1024))
        file.seek(0)  # Read a portion for analysis, then reset file pointer
        if file_type == "image/svg+xml":
            if raise_exception:
                raise ValidationError("SVG image upload is not allowed. Please use another image type.")
            return False
        return True
```

**Review by thenav56 — CHANGES_REQUESTED** [review]
@szabozoltan69 Look good.

I noticed the validation using `file_field` in validated_data. It is much easier to have a function `validate_<file_field>` instead, as it will handles both create and update, and we don't have to check if the field is provided by the client for different API methods.

**Comment by szabozoltan69** [comment]
@thenav56 I implemented the suggested changes. Really nicer this way! :-)

**Comment by szabozoltan69** [comment]
(Also tested it locally.)

**Review by thenav56 — APPROVED** [review]

**Pull request merged**